Resource agent for Microsoft Azure Scheduled Events

tniek opened the pull request:
This PR adds a resource agent ("AzEvents") for dealing with scheduled maintenance events inside Microsoft Azure VMs.

- It monitors the Azure Metadata API and, if events are found, tries to stop all or a subset of resources on the affected VM gracefully to migrate them to other nodes in the same cluster.

- Although application-agnostic, it is mainly targeted towards SAP applications deployed on a Pacemaker HA cluster.

- It should be implemented as clone resource with regular 'monitor' operation; intervals of 10 seconds are fine. (No other ops are implemented, since the RA does not handle any actual resources.)

Example via **crm node**:
```
primitive rsc_AzEvents ocf:heartbeat:AzEvents op monitor interval=10s
clone cln_AzEvents rsc_AzEvents
```

For more information on Azure Scheduled Events, please refer to:
https://docs.microsoft.com/en-us/azure/virtual-machines/linux/scheduled-events

Please let me know which changes are required to get this PR accepted. Thank you for your feedback in advance!

/cc @balaji-suse for following up on SAP-specific RAs
/cc @MSSedusch as author of other Azure-specific RAs
/cc @EricRadzikowskiMSFT as owner of Azure Scheduled Events

krig commented:
We'll need to have a conversation about how to include Python agents and what level of version compatibility we want to require. It looks like this agent is Python 3 only, right?

The current CI setup also assumes that the agents are all written in shell, so we need to update that to run something like pylint for Python agents...

Other than that, some general thoughts:

Currently, the agent is written to use `crm`, which isn't part of upstream pacemaker. One option would be to use `pcs` if that is installed instead, but it would probably be preferrable to use the builtin pacemaker commands to do what is currently done using `crm`.

The agent itself is pretty complex, so I'll need to take a closer look to see if I have any other concerns.

tniek commented:
Thanks @krig for taking a first look!

AzEvents is written in Python 2.7 and, as you noted, uses `crm`. That was a conscious decision, since I wanted to stay as close to the standard SLES 12.3 for SAP distribution as possible. (The RA should work for any application running on Pacemaker, but is clearly targeted towards SAP workloads.)

Let me know if you have any other questions. I appreciate your help!

krig commented:
So the agent being Python 2 only is going to be a problem, since both SLES and RHEL are going Python 3-only in a fairly short timespan. We will also need to update the build process to require Python in order to test the agent.

I would prefer it if it wasn't limited to SLES 12.3, but we could probably extend platform support once it is merged...

@oalbrigt Any comments on this, or ideas for how to make progress?

oalbrigt commented:
@krig I agree that it should work for both Python 2.7.x and 3.x for us to accept it, as those are the 2 main versions currently, and 2.x being removed from newer distros 3.x support should be a minimum.

I'll take care of the expected resource agent logging logic in Python, when I get my library released.

I've added some Python detection logic: https://github.com/ClusterLabs/resource-agents/pull/1170

So if they rebase on upstream/master, change the #!-line to `#!@PYTHON@ -tt`, rename the agent to it's name ending in ".in" and add it configure.ac (search for chmod +x) it'll be replaced during ./configure.

tniek commented:
@oalbrigt Thanks for your input; I've implemented each of your requested changes.

I'm not sure how exactly to test the configure changes, however, I've added my RA to configure.ac as requested.

Kindly let me know if you need anything else from my side to get this PR accepted.

tniek commented:
@krig @oalbrigt Just checking in if you need anything else from my side to get this PR accepted?

sameeksha91 commented:
@krig Hi Kristoffer, I work with tniek. Following up to understand if there are any other comments from you & team on improving this pull request. We'll be happy to make changes per your comments. Could you please let us know your comments or please approve this PR. Thank you! :)

krig commented:
Hi @sameeksha91, @tniek: There are a couple of things that should be fixed but I would be sort of OK with merging it without those (@oalbrigt is also free to chip in):

* Currently only supporting Python 2 as @oalbrigt pointed out above
* Currently using `crmsh` which is not included in RHEL

The best solution to the second problem would be to use only the core `pacemaker` command line tools rather than `crm`.

The main thing that has delayed things is that it is quite big and quite unusual, to be honest. I'll take a look tomorrow and see if there is anything else that I think should be done before merging.

krig commented:
Hmm, yes. So Python 3 compatibility does need to be fixed. :( That's why the CI check is failing at the moment. We need the agents to be able to return metadata in order to generate the man pages, and the azure-events agent fails here:

```
OCF_ROOT=/home/travis/build/ClusterLabs/resource-agents/heartbeat OCF_FUNCTIONS_DIR=/home/travis/build/ClusterLabs/resource-agents/heartbeat ./azure-events meta-data | xmllint --path /home/travis/build/ClusterLabs/resource-agents/heartbeat --noout --relaxng /home/travis/build/ClusterLabs/resource-agents/heartbeat/metadata.rng -
Traceback (most recent call last):
  File "./azure-events", line 13, in <module>
    from enum import Enum
ImportError: No module named enum
-:1: parser error : Document is empty
make[2]: *** [azure-events.check] Error 1
make[2]: Leaving directory `/home/travis/build/ClusterLabs/resource-agents/heartbeat'
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory `/home/travis/build/ClusterLabs/resource-agents'
make: *** [check] Error 2
```

Even if the rest of the agent needs Python 2, we do need it to be able to return metadata when running under Python 3.

oalbrigt commented:
+1 regarding Python 3-support and not using distro-specific tools.

oalbrigt commented on heartbeat/azure-events.in:
```diff
@@ -0,0 +1,893 @@
+
+import os, sys, time, subprocess
+import json
+import urllib, urllib2, socket
```
Regarding urllib here's an example of how it can be implemented for both Python 2 and 3:
https://github.com/ClusterLabs/fence-agents/blob/master/agents/gce/fence_gce.py#L8

sameeksha91 commented:
@krig @oalbrigt Thank you for your comments. They're helpful! I want to share that Azure-events targets a subset of customers - customers deploying SAP application (SAP HANA) on Microsoft Azure M-series Virtual Machines running SLES 11/12. On top of this, SAP HANA exclusively uses Python 2.7 (not 3). This mean that at this time with this pull request we are ready to serve this particular set of customers  well. Could I please urge you to accept this pull request as iteration 1 given that we are ready for target customers and we can sync again for iteration 2 with more of/rest of your comments. Please let us know your thoughts.

krig commented:
@sameeksha91 For Python 2/3 support, although SAP HANA uses Python 2.7, the fencing agents are expected to be able to run with the Python 3 interpreter, and this is how we run them in future service packs of SLES 12 as well as SLES 15. In addition to that, in order for the build pipeline to work with the agent it will also need to be runnable with Python 3. So that issue in particular is a major obstacle for inclusion for us.

As for the other comments, since the use of the `crm` command can be completely replaced with calls to `crm_attribute` which is available on all pacemaker-supported distributions, making that change would be helpful to anyone who wants to use the agent outside the context of SAP HANA.

diegoakechi commented:
@sameeksha91 and @tniek, if it helps, @krig and I can setup a video chat to clarify and try to speed up the adjustments needed to have this PR merged. Feel free to contact me via email for that.

sameeksha91 commented:
Thank you @krig and @diegoakechi! 

@diegoakechi @tniek does 8 am PST which is 5 pm CEST work for both?

tniek commented on heartbeat/AzEvents:
```diff
@@ -0,0 +1,893 @@
+		logging.debug("_sendMetadataRequest: begin; endpoint = %s, postData = %s" % (endpoint, postData))
+		logging.debug("_sendMetadataRequest: url = %s" % url)
+
+		req = urllib2.Request(url, postData)
```
Good idea - will include that. Thanks @harijayms!

krig commented:
Closing in favor of the updated PR #1248.
